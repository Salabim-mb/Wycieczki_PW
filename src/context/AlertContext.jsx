import React, {useState} from 'react';
import alertStyle from 'constants/alertOptions';

export const AlertContext = React.createContext({
	open: false,
	message: "",
	showAlert: () => {}
});

export const AlertProvider = (props) => {
	const [open, setOpen] = useState(false);
	const [message, setMessage] = useState("");
	const [severity, setSeverity] = useState(alertStyle.ERROR);

	const data = {
		open,
		message,
		severity,
		changeVisibility: (newOpen) => setOpen(newOpen),
		changeMessage: (newMessage, newSeverity=alertStyle.ERROR) => {
			setOpen(true);
			setMessage(newMessage);
			setSeverity(newSeverity);
		},
		showAlert: (newMessage, newSeverity) => {
			data.changeMessage(newMessage, newSeverity);
			data.changeVisibility(true);
		},
	};

	return <AlertContext.Provider value={data} {...props} />
};