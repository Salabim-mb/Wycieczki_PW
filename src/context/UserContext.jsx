import React, { useState } from 'react';
import Cookies from 'universal-cookie';
import { decryptCookie, encryptCookie } from 'utils/aesCiphrator';

const cookie = new Cookies();

export const UserContext = React.createContext({
	token: undefined,
	data: undefined,
	type: undefined,
	login: () => {},
	logout: () => {}
});

export const UserProvider = (props) => {
	const [token, setToken] = useState(
		decryptCookie(cookie.get("token")) || undefined
	);
	const [data, setData] = useState(
		decryptCookie(cookie.get("data")) || undefined
	);
	const [type, setType] = useState(
		decryptCookie(cookie.get("type")) || undefined
	)

	const user = {
		token,
		data,
		type,
		login: (newToken, newData, newType) => {
			setToken(newToken);
			setData(newData);
			setType(newType);
			cookie.set("token", encryptCookie(newToken), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
			cookie.set("data", encryptCookie(newData), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
			cookie.set("type", encryptCookie(newType), {
				path: "/",
				secure: process.env.REACT_APP_SECURE_COOKIES !== 'off'
			});
		},
		logout: () => {
			cookie.remove("token");
			cookie.remove("data");
			cookie.remove("type");
			setData(undefined);
			setToken(undefined);
			setType(undefined);
		},
	};

	return <UserContext.Provider value={user} {...props} />;
};