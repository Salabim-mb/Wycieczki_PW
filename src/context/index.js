export * from './AlertContext';
export * from './UserContext';
