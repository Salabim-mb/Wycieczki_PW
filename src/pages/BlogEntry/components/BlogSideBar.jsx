import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Breadcrumbs, Chip, Grid, Paper, Typography } from '@material-ui/core';
import NavigateNextIcon from '@material-ui/icons/NavigateNext';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
	wrapper: {
		maxWidth: '100%',
		padding: theme.spacing(2, 0)
	},
	sidebarBreadcrumbs: {
		padding: theme.spacing(2),
		backgroundColor: theme.palette.grey[200],
		margin: theme.spacing(4, 0, 3, 0)
	},
	sidebarSection: {
		marginTop: theme.spacing(3),
	},
}));

const BlogSideBar = ({pathParams, featuredPosts, setRedirect}) => {
	const classes = useStyles();

	return (
		<Grid item xs={12} md={4} className={classes.wrapper}>
			<Paper elevation={0} className={classes.sidebarBreadcrumbs}>
				<Typography variant="h6" component="h6">
					Nawigacja
				</Typography>
				<Breadcrumbs separator={<NavigateNextIcon fontSize="small" />} aria-label="breadcrumb">
					{pathParams.map((item) => (
						<Chip
							key={item.path}
							label={item.name}
							variant="outlined"
							onClick={() => setRedirect({action: true, destination: item.path})}
						/>
					))}
				</Breadcrumbs>
			</Paper>
			<Typography variant="h5" component="h6">
				Proponowane wpisy
			</Typography>
			Tu będą ze 3 tytuły proponowanych / losowych wpisów
			{featuredPosts}
		</Grid>
	);
};

BlogSideBar.propTypes = {
	pathParams: PropTypes.arrayOf(
		PropTypes.objectOf({
			name: PropTypes.string,
			path: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
		})
	).isRequired,
	featuredPosts: PropTypes.arrayOf(
		null // instancja posta
	).isRequired,
	setRedirect: PropTypes.func.isRequired
}

export default BlogSideBar;