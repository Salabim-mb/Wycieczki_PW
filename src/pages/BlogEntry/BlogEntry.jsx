import React, { useContext, useEffect, useRef, useState } from 'react';
import {Redirect, useParams} from 'react-router-dom';
import { Backdrop, CircularProgress, CssBaseline, Divider, Grid, Paper, Typography } from '@material-ui/core';
import { AlertContext } from 'context/AlertContext';
import { makeStyles } from '@material-ui/core/styles';
import { pathList } from 'constants/routes';
import BlogSideBar from './components/BlogSideBar';

const useStyles = makeStyles((theme) => ({
	mainGrid: {
		marginTop: theme.spacing(3),
	},
	entryBody: {
		...theme.typography.body2,
		padding: theme.spacing(3, 0),
	},
	backdrop: {
		zIndex: theme.zIndex.drawer + 1,
		color: '#fff',
	},
}));

const getEntryContent = async (id) => {
	const url = `/${id}/`;
	const headers = {
		"Content-Type": "application/json"
	}
	const res = await fetch(url, {
		headers,
		method: "GET"
	});

	if (res.status === 200) {
		return await res.json();
	} else {
		throw await res.json();
	}
};

const mapEntry = (data) => ({
	id: data.id,

});

const BlogEntry = () => {
	const {category, entryId} = useParams();
	const [redirect, setRedirect] = useState(false);
	const [entryData, setEntryData] = useState({});
	const [loading, setLoading] = useState(false);
	const alertC = useRef(useContext(AlertContext));

	useEffect(() => {
		const loadPost = async () => {
			setLoading(true);
			try {
				const res = await getEntryContent(entryId);
				setEntryData(mapEntry(res));
			} catch(ex) {
				alertC.current.showAlert(ex.message || "Coś poszło nie tak przy próbie pobierania posta.")
			} finally {
				setLoading(false);
			}
		};
		loadPost();
	});

	const mapDate = (date) => {
		const d = new Date(date);
		return `${d.getDate()} - ${d.getMonth() + 1} - ${d.getFullYear()}`;
	}

	const classes = useStyles();

	return (
		<Paper>
			{loading ? (
				<Backdrop open={loading} className={classes.backdrop}>
					<CircularProgress color="inherit" />
				</Backdrop>
			) : (
				<>
					<CssBaseline />
					<Grid container spacing={5} className={classes.mainGrid}>
						<Grid item xs={12} md={8}>
							<Typography variant="h6" gutterBottom>
								Utworzono: {entryData.title}
							</Typography>
							<Typography variant="caption" gutterBottom>
								{mapDate(entryData.date)}
							</Typography>
							<Divider />
							<div className={classes.entryBody}>
								{entryData.content}
							</div>
						</Grid>
						<BlogSideBar
							featuredPosts={[]}
							setRedirect={setRedirect}
							pathParams={[
								{name: "Strona główna", path: pathList.DASHBOARD.path},
								{name: category, path: pathList.BLOG_CATEGORIES.redirect(category)},
								{name: entryData.title, path: pathList.BLOG_ENTRY.redirect(category, entryId)}
							]}
						/>
					</Grid>
					{redirect?.action && <Redirect to={redirect?.destination || null} />}
				</>
			)}
		</Paper>
	)
};

export default BlogEntry;