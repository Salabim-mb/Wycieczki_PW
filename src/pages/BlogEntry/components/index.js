import BlogSideBar from './BlogSideBar';

export default BlogSideBar;