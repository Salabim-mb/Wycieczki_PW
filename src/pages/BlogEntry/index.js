import BlogEntry from 'pages/BlogEntry';

export default BlogEntry;