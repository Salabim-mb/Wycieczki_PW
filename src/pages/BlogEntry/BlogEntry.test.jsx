import React from 'react';
import { render, waitFor } from '@testing-library/react';
import {MemoryRouter} from 'react-router-dom';
import { AlertProvider } from 'context/AlertContext';
import BlogSideBar from './components/BlogSideBar';
import BlogEntry from './BlogEntry';

jest.mock("src/pages/BlogEntry/components");

describe("BlogEntry", () => {
	let apiFail;
	let entryData;
	// let params;
	
	beforeAll(() => {
		BlogSideBar.mockImplementation(() => <div>BlogSideBar</div>);
		global.fetch = jest.fn().mockImplementation(() => new Promise((resolve) => {
				if (apiFail) {
					resolve({status: 500, json: () => Promise.resolve({message: "not ok"})});
				} else {
					resolve({status: 200, json: () => Promise.resolve(entryData)});
				}
			}));
	});

	beforeEach(() => {
		apiFail = false;
		entryData = {
			id: 1,
			date: (new Date()).toISOString(),
			title: "TytuÅ‚",
			content: "lorem ipsum"
		};
		// params = {
		// 	category: "cat",
		// 	entryId: "123456"
		// };
	});

	it("should match snapshot", async () => {
		const {container, getByText} = render(
			<AlertProvider>
				<MemoryRouter>
					<BlogEntry />
				</MemoryRouter>
			</AlertProvider>
		);
		await waitFor(() => getByText(entryData.title));

		expect(container).toMatchSnapshot();

	});

	it("should display error alert on api fail", async () => {

	});
});