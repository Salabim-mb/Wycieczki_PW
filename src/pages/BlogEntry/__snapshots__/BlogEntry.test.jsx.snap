// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogEntry should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root makeStyles-backdrop-3"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
